### ISSUE TITLE
Panic when using coc.nvim :CocList command

### ISSUE THREAD

**Issue by juho-p** [vhakulinen/gnvim#92]
nvim version is 0.3.8. gnvim is compiled from current master.

Panic happens using following steps

- install coc.nvim plugin: https://github.com/neoclide/coc.nvim
- run :CocList -command

Terminal nvim works fine. I'm not sure exactly what coc.nvim is doing there.

Gnvim outputs following (including stacktrace). After typing `RUST_BACKTRACE=1 gnvim` command I immediately typed `CocList<cr>`. Crashes every time.

```
juho@juho-ThinkPad-P1:~$ RUST_BACKTRACE=1 gnvim
Not supported option set: arabicshape
Not supported option set: ambiwidth
Not supported option set: emoji
Not supported option set: guifontset
Not supported option set: guifontwide
Not supported option set: showtabline
Not supported option set: termguicolors
Not supported option set: ext_linegrid
Not supported option set: ext_hlstate
Not supported option set: ext_cmdline
Not supported option set: ext_popupmenu
Not supported option set: ext_tabline
Not supported option set: ext_wildmenu
Not supported option set: termguicolors
Not supported option set: ext_cmdline
Not supported option set: ext_popupmenu
Not supported option set: ext_tabline
Not supported option set: ext_wildmenu
Not supported option set: ext_cmdline
Not supported option set: ext_popupmenu
Not supported option set: ext_tabline
Not supported option set: ext_wildmenu
Not supported option set: ext_cmdline
Not supported option set: ext_popupmenu
Not supported option set: ext_tabline
Not supported option set: ext_wildmenu
Not supported option set: ext_cmdline
Not supported option set: ext_popupmenu
Not supported option set: ext_tabline
Not supported option set: ext_wildmenu
Failed to turn input event into nvim key (keyval: 65506)
Failed to turn input event into nvim key (keyval: 65506)
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:347:21
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: rust_begin_unwind
             at src/libstd/panicking.rs:308
   7: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   8: core::panicking::panic
             at src/libcore/panicking.rs:49
   9: gnvim::ui::grid::row::Rope::cell_at
  10: gnvim::ui::grid::grid::Grid::flush
  11: gnvim::ui::ui::UI::start::{{closure}}
  12: glib::main_context_channel::dispatch
  13: g_main_context_dispatch
  14: <unknown>
  15: g_main_context_iteration
  16: g_application_run
  17: <O as gio::application::ApplicationExtManual>::run
  18: gnvim::main
  19: std::rt::lang_start::{{closure}}
  20: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:293
  21: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:85
  22: std::rt::lang_start_internal
             at src/libstd/panicking.rs:272
             at src/libstd/panic.rs:394
             at src/libstd/rt.rs:48
  23: main
  24: __libc_start_main
  25: _start
```

**Comment by badosu** [comment]
Can't reproduce, can you provide your neovim and gtk versions?

**Comment by juho-p** [comment]
I tried this again with neovim  unstable (version 0.4.0 from here https://launchpad.net/~neovim-ppa/+archive/ubuntu/unstable), and it really does work! So this problem is only with stable version of nvim (0.3.8).

Gtk version is 3.22
```
~$ apt show libgtk-3-0 
Package: libgtk-3-0
Version: 3.22.30-1ubuntu4
```

**Comment by vhakulinen** [comment]
So you did resolve this issue?

**Comment by juho-p** [comment]
I resolved this on one computer by installing development version of nvim. The stable version 0.3.8 still crashes.

I did some debugging. The crash happens in row.rs: `let c = leaf.text.chars().nth(at - 1).unwrap();` where it tries to access character at index that is out of range (`leaf.text` is too short). That's called from grid.rs `flush` function

```
        // If cursor isn't blinking, drawn the inverted cell into the cursor's
        // cairo context.
        if ctx.cursor_blink_on == 0 {
            if let Some(row) = ctx.rows.get(ctx.cursor.0 as usize) {
                let cell = row.cell_at(ctx.cursor.1 as usize + 1);
```

After seeing that, I tried setting cursor to blink `set guicursor+=a:blinkon333` and that would make gnvim not to crash.

So, with "clean" config (only dein and coc.nvim installed), I get nvim 0.3.8 to crash just by running `:CocList` command. Hovever, when I run commands `:set guicursor+=a:blinkon333` followed by `CocList` gnvim does *not* crash.

I don't know if this is really a problem in nvim 0.3.8 or gnvim. nvim 0.4.0 seems to fix the problem. Maybe older nvim reports cursor position wrong? (I'm just assuming here without knowing the code that `ctx.cursor` is the current `(row,column)` position of nvim cursor)

**Comment by dbarnett** [comment]
I also ran into this using vim-test to run a vader.vim test file in Neomake. I installed https://github.com/janko/vim-test, ran `gnvim tests/completion.vader`, then executed `:TestFile`, and after a second gnvim crashed with an error like the one you posted. I confirmed that doing `set guicursor+=a:blinkon333` before executing `:TestFile` does prevent the crash.

Upgrading to unstable neovim or forcing settings might avoid the crash, but it very much looks like a bug in gnvim and not neovim's problem. Can it be fixed?

**Comment by TyberiusPrime** [comment]
I can report that the same issue (and the same workaround) occurs with vim-peekaboo. As soon as the peekaboo window disappears, panic.

**Comment by smolck** [comment]
> Upgrading to unstable neovim or forcing settings might avoid the crash, but it very much looks like a bug in gnvim and not just neovim's problem. Can it be fixed?

@dbarnett What makes you say that?

**Comment by dbarnett** [comment]
Well, are neovim bugs expected to instakill gnvim with a mysterious backtrace pointing to a line of gnvim code? Just from how cryptic the error is it looked like a scenario gnvim never accounted for vs. saying "nvim did something unexpected that we can't recover from". Even if it is a straight-up neovim bug, there'd still be room to improve assertions and those user errors in gnvim, no?

**Comment by smolck** [comment]
> Even if it is a straight-up neovim bug, there'd still be room to improve assertions and those user errors in gnvim, no?

For sure, GNvim errors could definitely be improved. 

> Well, are neovim bugs expected to instakill gnvim with a mysterious backtrace pointing to a line of gnvim code? Just from how cryptic the error is it looked like a scenario gnvim never accounted for vs. saying "nvim did something unexpected that we can't recover from". 

The thing is, if Neovim sent a redraw message that it never should have sent due to a bug, GNvim wouldn’t be expecting it and could thus crash. Here, GNvim never accounted for it since Neovim never should have sent an invalid message (afaik. @vhakulinen would know more about this than I do).

If crashes like these don’t happen in newer Neovim versions, I don’t think we should patch GNvim for them unless we have to (or it really is a legitimate GNvim bug). And (@bfredl pointed this out to me on Neovim’s gitter), does #103 (which should fix this and similar crashes) make sense logically speaking? I’m not really sure.

It just seems a little suspicious to me that these crashes don’t appear to happen with newer Neovim versions.

As for Neovim bugs instakilling Gnvim, we hope that Neovim bugs don’t do that ;) But bugs like that we can submit issues for upstream, and they many times get fixed right away (the Neovim team is incredible).

**Comment by dbarnett** [comment]
Gotcha, I agree with your reasoning and appreciate you took the time to spell out details I didn't have. =)

For this `called `Option::unwrap()` on a `None` value`, if you'd be interested in some kind of handling a little further up the stack to call out the problem more along the lines of what you just described, I think that'd be a huge improvement. For instance just wrapping the error with "neovim did something unexpected. Exiting. See stack trace for details" would help people narrow down what went wrong and where to search for known issues.

**Comment by smolck** [comment]
> Gotcha, I agree with your reasoning and appreciate you took the time to spell out details I didn't have. =)

No problem! Thank you for taking the time to read and consider said details ;)

> For this `called `Option::unwrap()`on a`None` value`, if you'd be interested in some kind of handling a little further up the stack to call out the problem more along the lines of what you just described, I think that'd be a huge improvement. For instance just wrapping the error with "neovim did something unexpected. Exiting. See stack trace for details" would help people narrow down what went wrong and where to search for known issues.

I 100% agree with you. As a user seeing “thread 'main' panicked at 'called `Option::unwrap()` on a `None` value'” is confusing and can be misleading. Would you like to submit a PR for this? I’m happy to help out and/or answer any questions you may have.

**Comment by smolck** [comment]
I'm a bit unsure where we'd want to do this though. This issue's occurring at this line: https://github.com/vhakulinen/gnvim/blob/7b89a6de910252acf86c5035155b9df63fd46f31/src/ui/grid/row.rs#L238

However, I don't necessarily think we should just change that to `leaf.text.chars().nth(at - 1).expect("Neovim did something unexpected...");`. We should look at places further up the stack (as you said) where we assume things, e.g. places where call `.unwrap()`. But we also should avoid placing `.expect("...")` everywhere we see `.unwrap()`.

TL;DR We'll need to strike a balance between having better error messages in places where Neovim bugs could potentially show themselves and littering the code with unnecessary `.expect()`'s.

**Comment by vhakulinen** [comment]
Is this still happening on nvim 0.4?

**Comment by vhakulinen** [comment]
Can you test with nvim 0.4 and the latest master of gnvim?

**Comment by juho-p** [comment]
I just tried the master version of gnvim. This is not happening on nvim 0.4.2 that is the most recent stable, though latest Ubuntu, that I use, still has 0.3.8. Anyway, it's perfectly fine for me if you prefer to close this issue.

Though now I have (fully unrelated to this) weird issue with cursor getting invisible that's not happening in command line neovim... I'll try to figure out later if some of the plugins I use are causing this.

**Comment by vhakulinen** [comment]
For anyone following this issue (or coming across this), try with latest gnvim master (currently https://github.com/vhakulinen/gnvim/commit/16b465b967533386f380a1696b9f7c7b9b89e77e) and nvim 0.4.2.
